src/core/SablierLockup.sol

```diff
@@ -0,0 +1,508 @@
+import { Lockup, LockupDynamic, LockupLinear, LockupTranched } from "./types/DataTypes.sol";
+
+/*
+
+███████╗ █████╗ ██████╗ ██╗     ██╗███████╗██████╗     ██╗      ██████╗  ██████╗██╗  ██╗██╗   ██╗██████╗
+██╔════╝██╔══██╗██╔══██╗██║     ██║██╔════╝██╔══██╗    ██║     ██╔═══██╗██╔════╝██║ ██╔╝██║   ██║██╔══██╗
+███████╗███████║██████╔╝██║     ██║█████╗  ██████╔╝    ██║     ██║   ██║██║     █████╔╝ ██║   ██║██████╔╝
+╚════██║██╔══██║██╔══██╗██║     ██║██╔══╝  ██╔══██╗    ██║     ██║   ██║██║     ██╔═██╗ ██║   ██║██╔═══╝
+███████║██║  ██║██████╔╝███████╗██║███████╗██║  ██║    ███████╗╚██████╔╝╚██████╗██║  ██╗╚██████╔╝██║
+╚══════╝╚═╝  ╚═╝╚═════╝ ╚══════╝╚═╝╚══════╝╚═╝  ╚═╝    ╚══════╝ ╚═════╝  ╚═════╝╚═╝  ╚═╝ ╚═════╝ ╚═╝
+
+*/
+
+/// @title SablierLockup
+/// @notice See the documentation in {ISablierLockup}.
+contract SablierLockup is ISablierLockup, SablierLockupBase {
+    using SafeERC20 for IERC20;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                  STATE VARIABLES
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierLockup
+    uint256 public immutable override MAX_COUNT;
+
+    /// @dev Cliff timestamp mapped by stream IDs. This is used in Lockup Linear models.
+    mapping(uint256 streamId => uint40 cliff) internal _cliffs;
+
+    /// @dev Stream segments mapped by stream IDs. This is used in Lockup Dynamic models.
+    mapping(uint256 streamId => LockupDynamic.Segment[] segments) internal _segments;
+
+    /// @dev Stream tranches mapped by stream IDs. This is used in Lockup Tranched models.
+    mapping(uint256 streamId => LockupTranched.Tranche[] tranches) internal _tranches;
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                     CONSTRUCTOR
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @param initialAdmin The address of the initial contract admin.
+    /// @param initialNFTDescriptor The address of the NFT descriptor contract.
+    /// @param maxCount The maximum number of segments and tranched allowed in lockup dynamic and lockup tranched
+    /// streams, respectively.
+    constructor(
+        address initialAdmin,
+        ILockupNFTDescriptor initialNFTDescriptor,
+        uint256 maxCount
+    )
+        ERC721("Sablier Lockup NFT", "SAB-LOCKUP")
+        SablierLockupBase(initialAdmin, initialNFTDescriptor)
+    {
+        MAX_COUNT = maxCount;
+        nextStreamId = 1;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                           USER-FACING CONSTANT FUNCTIONS
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc ISablierLockup
+    function getCliff(uint256 streamId) external view override notNull(streamId) returns (uint40 cliff) {
```

Comment:
I thought about it. But then in `getTimestamps`, we return `cliff`. So it did not make sense for me to revert on `getCliff` but return `cliff `in `getTimestamps`. Otherwise I agree with you. Whats your thought on this?

Reply:
we decided not to include the cliff in timestamps, so we should revert

Reply:
Agree.